File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;
class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        int player = stages.length;

        int[] step = new int[502];
        List<Double> list = new ArrayList<>();
        List<Double> index = new ArrayList<>();

        for(int i = 0; i < stages.length; i++){
            step[stages[i]]++;
        }

        for(int j = 0; j < N; j++){
            if(player != 0){
                list.add((1.0 * step[j + 1]) / player);
                index.add(list.get(j));
            }else {
                list.add(0.0);
                index.add(0.0);
            }
            player -= step[j + 1];
        }

        Collections.sort(list);

        for(int k = 0; k < N; k++){
            answer[k] = index.indexOf(list.get((N - 1) - k)) + 1;
            index.set(index.indexOf(list.get((N - 1) - k)), -1.0);

            list.set((N - 1) - k, -1.0);
        }
        
        return answer;
    }
}

File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    p = len(stages)
    fail = [0] * (N + 2)
    result = {}

    for i in stages:
        fail[i] += 1

    for j in range(1, N + 1):
        if p == 0:
            result[j] = 0
        else:
            result[j] = fail[j] / p
        p -= fail[j]
        
    for l, n in sorted(result.items(), key = lambda x : (-x[1], x[0])):
        answer.append(l)
    
    return answer